### PR TITLE
added support for client_credentials oauthentication

### PR DESCRIPTION
**Pull request by ghost**
Helpfull probably for others too.

### CODE REVIEW

**Comment by trekawek** [comment]
Thanks for the PR! I'll review it and merge it soon.

File: src/main/java/com/adobe/analytics/client/auth/ClientCredentialsOAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.ProtocolException;
+import java.net.URL;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+public class ClientCredentialsOAuthenticator extends OAuthenticator {
+    private static final String OAUTH_URL = "https://api.omniture.com/token";
```
Comment (trekawek) [inline src/main/java/com/adobe/analytics/client/auth/ClientCredentialsOAuthenticator.java]: Please use tabs in all files to make the patch consistent with the codebase.

File: src/main/java/com/adobe/analytics/client/auth/ClientCredentialsOAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    private void setBasicAuth(HttpURLConnection conn) {
+        conn.setRequestProperty("Authorization", getAuthHeader());
+    }
+
+    private void writePostParameter(HttpURLConnection conn) throws IOException {
+        String urlParameters = "grant_type=client_credentials";
+        byte[] postData = urlParameters.getBytes(StandardCharsets.UTF_8);
+        int postDataLength = postData.length;
+        conn.setRequestProperty("Content-Length", Integer.toString(postDataLength));
+        try (DataOutputStream wr = new DataOutputStream(conn.getOutputStream())) {
```
Comment (trekawek) [inline src/main/java/com/adobe/analytics/client/auth/ClientCredentialsOAuthenticator.java]: I'm not sure if the DataOutputStream is the right choice here. Maybe we can use the [`IOUtils.write`](https://github.com/Adobe-Marketing-Cloud/analytics-java-library/blob/master/src/main/java/com/adobe/analytics/client/AnalyticsClient.java#L57) method as in the AnalyticsClient?

File: src/main/java/com/adobe/analytics/client/auth/OAuthenticator.java [hunk 3]
```diff
@@ -1,66 +1,46 @@
...
-				.signWith(SignatureAlgorithm.HS256, privateKey).compact();
-		final URL url = new URL(String.format(OAUTH_URL, endpoint, clientId, jwt));
-		final HttpURLConnection connection = (HttpURLConnection) url.openConnection();
-		final JsonObject response = JsonUtil.GSON.fromJson(ConnectionUtil.readResponse(connection),
-				JsonObject.class);
-		accessToken = response.get("access_token").getAsString();
-		expires = Calendar.getInstance();
-		expires.add(Calendar.SECOND, response.get("expires_in").getAsInt());
-	}
+    public void setAccessToken(String accessToken) {
```
Comment (trekawek) [inline src/main/java/com/adobe/analytics/client/auth/OAuthenticator.java]: This method should be private, so no one from the outside can change the internal state of the authenticator.

File: src/main/java/com/adobe/analytics/client/auth/OAuthenticator.java [hunk 4]
```diff
@@ -1,66 +1,46 @@
...
+    }
 
-	@Override
-	public void authenticate(HttpURLConnection connection) throws JsonSyntaxException, IOException {
-		if (!isTokenValid()) {
-			login();
-		}
-		connection.addRequestProperty("Authorization", String.format("Bearer %s", accessToken));
-	}
+    public void setExpires(Calendar expires) {
```
Comment (trekawek) [inline src/main/java/com/adobe/analytics/client/auth/OAuthenticator.java]: This method should be private as well.

**Comment by trekawek** [comment]
@ByteFader, I added a few minor comments. Could you apply them before I merge the PR? Thanks!

**Comment by ghost** [comment]
Fixed your PR remarks

**Pull request merged**